cli/cli.go

```diff
@@ -0,0 +1,168 @@
+// Package cli provides boilerplate code required to generate plugin executable.
+// At high level it performs following tasks
+// 1. Validate command arguments
+// 2. Read and unmarshal input
+// 3. Execute relevant plugin functions
+// 4. marshals output
+
+package cli
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"reflect"
+
+	"github.com/notaryproject/notation-plugin-framework-go/internal/slices"
+	"github.com/notaryproject/notation-plugin-framework-go/log"
+	"github.com/notaryproject/notation-plugin-framework-go/plugin"
+)
+
+type Cli struct {
+	syntax string
+	pl     plugin.Plugin
+}
+
+// New creates a new Cli using given plugin
+func New(pl plugin.Plugin) *Cli {
+	return &Cli{pl: pl}
+}
+
+// Execute is main controller that reads/validates commands, parses input, executes relevant plugin functions
+// and returns corresponding output.
+func (c Cli) Execute(ctx context.Context, args []string) {
+	c.validateArgs(ctx, args)
+
+	rescueStdOut := deferStdout()
+	command := args[1]
```

Comment:
What if `len(args) == 0`?

Reply:
Its not possible because we are calling `c.validateArgs(ctx, args)` at line 35 which performs following check 
```
if !(len(args) == 2 && slices.Contains(getValidArgs(md), args[1])) {
    deliverError(fmt.Sprintf("Invalid command, valid choices are %s %s", c.name, getValidArgsString(md)))
}
```